services/QuillLMS/engines/evidence/app/models/evidence/research/gen_ai/step_connection.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: evidence_research_gen_ai_step_connections
+#
+#  id             :bigint           not null, primary key
+#  condition      :string
+#  created_at     :datetime         not null
+#  updated_at     :datetime         not null
+#  source_step_id :bigint           not null
+#  target_step_id :bigint           not null
+#
+module Evidence
+  module Research
+    module GenAI
+      class StepConnection < ApplicationRecord
+        belongs_to :source_step, class_name: 'WorkflowStep'
+        belongs_to :target_step, class_name: 'WorkflowStep'
+        validates :source_step_id, :target_step_id, presence: true
+        validate :steps_belong_to_same_workflow
+        validate :no_circular_references
+
+        def self.connected_to_steps(step_ids)
+          where(source_step_id: step_ids).or(where(target_step_id: step_ids))
+        end
+
+        private def steps_belong_to_same_workflow
+          return unless source_step && target_step
+          return if source_step.workflow_id == target_step.workflow_id
+
+          errors.add(:base, 'Source and target steps must belong to the same workflow')
+        end
+
+        private def no_circular_references
```

Comment:
Probably worth expanding this logic for multi-step circular references (either here or somewhere else), e.g. 1->2->1

Reply:
When building this out, I first [spiked](https://github.com/empirical-org/Empirical-Core/commit/12efb792f8f019bab7fd8815368eefb1071ff915) out a proof-of-concept to accommodate all four types outlined in https://www.anthropic.com/research/building-effective-agents.

The EvaluatorOptimizer node will have two components, an evaluator and an optimizer, as well as the evaluation condition which determines if the node should be called on itself again.  This permits a looping operation within the node, but the overall graph structure remains acyclic, which makes various graph operations (e.g. traversal) easier.